### DISCUSSION TITLE
How to add new encryption algorithms support for strongswan

### DISCUSSION THREAD

**Discussion by lijh8** [strongswan/strongswan#552] [General]
Hi experts,

I want to add Chinese encryption algorithms like SM2, SM3, SM4 which are already implemented in openssl 1.1.1 in 2018 to strongwan.

1. What's the right way to add new algorithms support for strongswan?
2. Can both the IKE negotiation and the actual encrypted data transfer use the same algorithm like Chinese SM2.

**Comment by tobiasbrunner** [comment]
I guess you need to patch the _openssl_ plugin so these algorithms are made available to the daemon. Since there are no identifiers for IKEv2/Psec defined for these algorithms, you need to use some from the private use range, which might cause interoperability issues (you can then add keywords for these algorithms via `proposal_keywords_t::register_token()` and for IPsec you can add mappings for the kernel via `kernel_interface_t::register_algorithm`).

> Can both the IKE negotiation and the actual encrypted data transfer use the same algorithm like Chinese SM2.

Depends on whether the Linux kernel supports these algorithms (might need patches there, in particular for the XFRM API in xfrm_algo, where I don't see any mention of them). Alternatively, you could use the [kernel-libipsec plugin](https://wiki.strongswan.org/projects/strongswan/wiki/kernel-libipsec) for IPsec in userland i.e. strongSwan (not really recommended, though).

**Reply by lijh8** [reply]
Thank you very much, Tobias
